### ISSUE TITLE
What is really the minimum SDK for the app?

### ISSUE THREAD

**Issue by applikationsprogramvara** [rive-app/rive-android#5]
I've managed to start the app on Android with SDK >= 23 as it is defined in `build.gradle`.

Then I saw the commit [added porter duff mode for android sdk <23](https://github.com/rive-app/rive-android/commit/6dec87d4ae508ea07545a4ee326ee510d4b5d0f9) and wondered if the lower SDKs are supported.

The gradle sync and build went just OK with `minSdkVersion 21`.

But on the device I have the error:
```
java.lang.UnsatisfiedLinkError: dlopen failed: 
     empty/missing DT_HASH in "libjnirivebridge.so" (built with --hash-style=gnu?)
at java.lang.Runtime.loadLibrary(Runtime.java:371)
at java.lang.System.loadLibrary(System.java:989)
at app.rive.runtime.kotlin.File.<clinit>(File.kt:11)
at app.rive.runtime.example.MainActivity.onCreate(MainActivity.kt:19)
```
Which happens directly on opening the file:

``` kotlin
var file = File(
    getResources().openRawResource(R.raw.off_road_car_blog).readBytes()
)
```

What is really the minimum SDK? 
What should be done to get it done? Rebuild `libjnirivebridge.so` with other parameters?

**Comment by umberto-sonnino** [comment]
Currently, we're targeting SDK 23, as you found in build.gradle, and they all share libraries have been built for that SDK.
If you want to try building `rive-android` for lower SDKs, take a look at the build scripts under the [/cpp](/cpp) folder, you might to start from [build.rive.for.sh](/cpp/build.rive.for.sh). 
Make sure you have the submodules checked out and [premake5](https://premake.github.io/download.html) installed, as compilation requires it..!

**Comment by applikationsprogramvara** [comment]
Thank you for your hint! Here are some results and further thoughts regarding the topic:

I've rebuilt the libraries using `build.rive.for.sh` and got a new error:
```
java.lang.UnsatisfiedLinkError: dlopen failed: 
  cannot locate symbol "__register_atfork" referenced by "libjnirivebridge.so"...
```
Which is explained [here](https://android.googlesource.com/platform/bionic/+/master/android-changes-for-ndk-developers.md#available-in-api-level-23).

After changing to `API=21` in `build.rive.for.sh` and the new rebuild, I stumbled on the error:
```
app.rive.runtime.example A/art: sart/runtime/indirect_reference_table.cc:113] 
JNI ERROR (app bug): local reference table overflow (max=512)
 local reference table dump:
   Last 10 entries (of 512):
       511: 0x731000d8 android.graphics.PorterDuff$Mode
       510: 0x12d5e780 android.graphics.PorterDuffXfermode
       509: 0x12d5e780 android.graphics.PorterDuffXfermode
```
As I understood this issue is coming from multiple `env->NewGlobalRef`, `env->GetStaticFieldID` etc. e.g. in `jni_refs.cpp` The total number of references seems for me way less than 512, although Android thinks another way.

I've found recommendation of removing references using `env->DeleteLocalRef`, e.g. [here](https://stackoverflow.com/questions/11109497/android-jni-error-app-bug-local-reference-table-overflow-max-512). It looks like in higher SDKs the number of references has increased from 512 to 65535, which is mentioned e.g. [here](https://stackoverflow.com/questions/42596694/how-to-fix-error-jni-error-app-bug-global-reference-table-overflow-max-6553).

At same time all mentioned above versions of the library are working all right on the devices with higher SDK.

What do you think about fixing this error? I am considering if it makes sense to invest more time for support devices with SDK < 23 if their share is 10% and decreasing.

**Comment by umberto-sonnino** [comment]
This is great! Thanks for taking the time to look into this!
There are also a few performance trade-offs to consider - if we delete those references we'll have to have to grab them at runtime - if we're doing this to target older devices, the hit might be noticeable. We'll have to discuss it to understand what the effort would be, and if it's worth doing.

Are you currently just experimenting with this runtime? Or are you trying to target a specific API level?

**Comment by applikationsprogramvara** [comment]
I have a new idea! I've just tried the app on a device, which has exactly Android 6.0.1 (SDK 23) and got the same error `JNI ERROR (app bug): local reference table overflow (max=512)`. This device though is amrv7 (32-bit). 

My idea is that `local reference table` is limited by 512 on 32-bit devices and by 65535 on 64-bit devices. 

If this idea is right, then this is an additional argument to handle JNI references carefully. I can't find any actual and reliable statistics for architecture usage globally on Android, but from my Google Play data I would estimate usage of 32-bit architecture around 25%, which is way more 5-10% of devices with SDK < 23. 

Also I found some in [JNI tips](https://developer.android.com/training/articles/perf-jni.html#local-and-global-references), that only 16 (??) reference are assumed normal:
> The implementation is only required to reserve slots for 16 local references, so if you need more than that you should either delete as you go or use `EnsureLocalCapacity`/`PushLocalFrame` to reserve more.

Sorry, I don't want to interfere your plans. You are doing the great job! I would say, if you think to do the fix now is difficult, then it is better to produce a working version "with limitations" earlier, than to infinitely build up backward compatibility.
So I'm not targeting any specific SDK. I would like to bring a little bit of "enjoyment" to the users by adding some short animations instead of dull dialogs or waiting wheels. So if it would work for 3/4 of the users, it would be also all right.

_BTW, I tried to use a riv-file exported from preview.rive.app in the android app and it crashed with another error
I know this is off topic, so tell me where it's better to address it, e.g. in Discord or open an another issue._

**Comment by applikationsprogramvara** [comment]
I have some discoveries, but I need some time to write them down. So I will send a note later.

**Comment by applikationsprogramvara** [comment]
Here are my further discoveries (partially off topic):

In parallel with our conversation I tried to play a riv-file downloaded from preview.rive.app and it actually crashed all the times although it was a plain circle. I asked in the Discord-channel and Guido recommended me to check if I use the latest version of the code.

So I really used the code of the master branch, assuming, that it is the latest and the most stable code. I checked and discovered, that there is one another branch updateRiveCpp, posted 5 days ago. So I checked it out and with new so-libraries the riv-files from preview.rive.app played just fine. (As I am writing these lines, I see, that branch updateRiveCpp is already merged into master, which is good)

On the next step I tried to rebuild the library for SDK 21 and it worked out just well.

Here are results of my tests:

- The following repository examples from [raw](https://github.com/rive-app/rive-android/tree/master/app/src/main/res/raw) folder cannot be opened anywhere, so I would suggest to remove them from the repository:
  - `iron_giant.riv`
  - `little_bird_learns_to_fly.riv`
  - `off_road_car_demo.riv`
- Files can be opened on any SDK down to 21 just well:
  - any file downloaded from the editor https://preview.rive.app/
  - `blend_testing.riv` from `raw`-folder
- The last file `off_road_car_blog.riv` is thought a little bit controversial:
  - it cannot be opened on lower SDKs 21 or 23, presumably because 32-bit, it ends up with `JNI ERROR (app bug): local reference table overflow (max=512)` as I mentioned above
  - it can be opened just fine on higher SDKs: confirmed on 28 and 29, on 64 bit devices
  - I don't know what is the real reason for it and I can't reproduce the technics used in the file, e.g. bones constraints are not yet displayed in Rive 2

From all these experiments I have the following conclusions:
- The app is working on SDK 21
- Thought there are some limitations - it is difficult to comprehend where they are coming from - maybe you can formalize that

If the discovered limitations stay (from practical point of view) I would wish to detect beforehand if the riv-file crashes on the device, so I can show to the user some alternative UI and avoid the crash.

Could you please comment it somehow and the issue can be closed.

**Comment by applikationsprogramvara** [comment]
Actually I stumbled on the same issue again. My newly exported riv-files are stopped to be opened by the Android app. I used your branch `updates_cleanup` to retrieve the new libraries and it worked just OK. By the time I'm writing this part is merged to the master, which is great!

Could you maybe open wiki for this repository? I would write some of my notes regarding rebuilding libraries for lower SDKs and maybe something else in the future. They could be useful for others.

**Comment by umberto-sonnino** [comment]
> Could you maybe open wiki for this repository? I would write some of my notes regarding rebuilding libraries for lower SDKs and maybe something else in the future. They could be useful for others.

While opening the Wiki is certainly a good idea, we're still a small team, and maintaining, reviewing, and keeping it up to date might be a bit too much for us.

If you want to contribute to our documentation, though, feel free to open up PRs for our README!

**Comment by applikationsprogramvara** [comment]
Do you have any updates regarding the subject and the reliability of the runtime for Android. I'm considering the push an app with an animation to production. 

It is OK, if animation is played only on the part of all the devices. The main criteria is **reliability** - there should be no crashes.

What kind of check should I put before starting the animation to avoid crashes? 

_For the beginning we can consider SDK >= 23 and platform is 64bit?_

**Comment by umberto-sonnino** [comment]
> For the beginning we can consider SDK >= 23 and platform is 64bit?

We're still at an early stage with the API and we're maturing it - that means that we'll also try to run more thorough tests to give more reliable answers about compatibility and stability.  
But, from what we've seen, and from your reports (which have been greatly appreciated, btw!), it looks like targeting these should be safe - these are the set of platforms we're aiming to support in the future in any case.

If you run into any trouble, feel free to contact us, we're happy to provide whatever support we can!
